CDK migration to multi-module Terraform (#146)
dominodatalab/cdk-cf-eks: convert/README.md

* convert.py - The core migration script
* README.md - This file
* cdk_tf/ - Terraform to take over CDK infrastructure that will not be imported onto the terraform-aws-eks module.
* cloudformation-only/ - Terraform to create a special IAM role to remove the CDK CloudFormation stack

## Migration steps

To convert the CDK stack to terraform, we will...

1. Generate the terraform modules structure.
2. Generate input variables into Terraform
3. Generate import blocks into Terraform
4. Initialize the Terraform directory.
5. Evaluate the Terraform plans.
6. Apply the Terraform plans.
7. Scale down the old auto-scaling groups.
8. Validate Domino functionality.
9. Delete unneeded AWS resources.
10. Delete the old CloudFormation stack.

### Prerequisites
* terraform version >= 1.5.0
* python 3.11
* jq

> * [hcledit](https://github.com/minamijoyo/hcledit)

Author: Released on bash 4 https://tldp.org/LDP/abs/html/bashver4.html. I think its fair to require a version released 14 years ago.

> * [hcledit](https://github.com/minamijoyo/hcledit)

Reviewer: Apple should really suck it up on the gpl...

* cloudformation-only/ - Terraform to create a special IAM role to remove the CDK CloudFormation stack

## Migration steps

To convert the CDK stack to terraform, we will...

1. Generate the terraform modules structure.
2. Generate input variables into Terraform
3. Generate import blocks into Terraform
4. Initialize the Terraform directory.
5. Evaluate the Terraform plans.
6. Apply the Terraform plans.
7. Scale down the old auto-scaling groups.
8. Validate Domino functionality.
9. Delete unneeded AWS resources.
10. Delete the old CloudFormation stack.

### Prerequisites
* terraform version >= 1.5.0
* python >=3.9
* jq
* [hcledit](https://github.com/minamijoyo/hcledit)
* Bash >= 4

### convert.py usage

    usage: convert.py [-h] {create-tfvars,resource-map,set-imports,delete-stack,print-stack} ...

    terraform eks module importer

    options:
      -h, --help            show this help message and exit

    commands:
      {create-tfvars,resource-map,set-imports,delete-stack,print-stack}
	create-tfvars       Generate tfvars
	resource-map        Create resource map for customization/debugging of set-imports command (optional step)

> 	set-imports         Get terraform import commands

Reviewer: Should get the updated help text here.

Validate the the requirements are installed and the expected version.

    ./convert.py check-requirements

### Set environment variables
Set `AWS_REGION`, `DEPLOY_ID` and `MOD_VERSION` environment variables with appropriate values.

* `AWS_REGION`  AWS region for the CloudFormation stack.
* `DEPLOY_ID` : Name of the main CloudFormation stack.
* `MOD_VERSION`: Release tag for [terraform-aws-eks](https://github.com/dominodatalab/terraform-aws-eks/releases) in the form `vX.Y.Z` (using `v3.0.0` as an example).

Command:

    export AWS_REGION='us-east-1' DEPLOY_ID='my-main-stack-name' MOD_VERSION='v3.0.0'

### Set Config values
Write environment variables' values to config file.

Command:

     envsubst < config.tpl | tee config.yaml

### Setup the terraform modules.
The following command will create a directory named after variable `$DEPLOY_ID` where it will Initialize the necessary terraform configuration.
It will also copy over `cdk_tf` under the  `$DEPLOY_ID/terraform` directory to centralize the terraform configuration.

Command:

    ./convert.py setup-tf-modules

### Create terraform variables

First, we must create the variables going into our Terraform config.

> * The following command will look at the current CDK stack and autopropagate the Terraform variables onto the appropriate modules. It will generate multiple files with extension `*.tfvars.json` under `$DEPLOY_ID/terraform`.

Reviewer: nit edit, apologies.

```suggestion
* The following command will look at the current CDK stack and autopropagate the Terraform variables for the appropriate modules by generating multiple files with extension `*.tfvars.json` under `$DEPLOY_ID/terraform`.
```

we used an existing keypair, while the terraform module creates a keypair for
you. The local key does not need to be the one you used with CDK. The new
autoscaling group nodes and bastion node will be provisioned with the new key.

### Get AWS resources to import

Then, we'll use `convert.py` to generate the Terraform import blocks to import our AWS resources. It will generate a file called `imports.tf` under each of the terraform modules located at:

* `$DEPLOY_ID/terraform/cdk_tf/`
* `$DEPLOY_ID/terraform/infra/`
* `$DEPLOY_ID/terraform/cluster/`
* `$DEPLOY_ID/terraform/nodes/`

Command:

    ./convert.py set-imports --region us-east-1 --stack-name mystack

### Review and Configure Node Groups

This conversion process will create a tvfars file with the default nodegroups for the terraform-aws-eks module. The variables file which corresponds to the nodes configuration is located at `$DEPLOY_ID/terraform/nodes.tfvars.json`
Please review the instructions for that module, and configure the nodegroups as desired. Your old nodegroups will remain functional after running the conversion (until the "clean-stack" step).

### Evaluate the Terraform plan

> Change onto the `DEPLOY_ID` directory:

Reviewer: ```suggestion
Change into the `DEPLOY_ID` directory:
```
